### PR TITLE
FC + elementwise_add (Residual connection)

### PR DESCRIPTION
**Pull request by Silv3S**
### PR types
Performance optimization

### PR changes
OPs

### Describe
- Fuse FC with elementwise_add so residual data is added inside FC operator,
- add unit tests with multiple cases for this fuse pass,

### Benchmarks
Test were conducted with oneDNN 2.5.4 , `fc_mkldnn_pass` and `fc_act_mkldnn_pass` were enabled:
- VIT-OCR is performing 4% faster (12 fc+residual fusions),
- BERT is performing 8% faster (24 fc+residual fusions).

### CODE REVIEW

File: paddle/fluid/operators/fc_op.cc [hunk 1]
```diff
@@ -137,6 +137,10 @@ class FCOpMaker : public framework::OpProtoAndCheckerMaker {
     AddInput("W", "(Tensor), The weight fc op with shape (I, O).");
     AddInput("Bias", "(Tensor, optional) Bias vector with shape (1 x O")
         .AsDispensable();
+    AddInput("ResidualData",
+             "(Tensor) Residual data which will be added to fc output")
+        .AsDispensable()
+        .AsExtra();
```
Comment (tsocha) [inline paddle/fluid/operators/fc_op.cc]: Is it allowed to add new attributes into existing operators?

File: paddle/fluid/operators/fc_op.cc [hunk 2]
```diff
@@ -179,10 +183,20 @@ class FCOpMaker : public framework::OpProtoAndCheckerMaker {
...
     AddAttr<bool>("force_fp32_output",
                   "(bool, default false) Force INT8 kernel output FP32, only "
                   "used in MKL-DNN INT8")
         .SetDefault(false);
+    AddAttr<bool>("fuse_residual_connection",
+                  "(bool, default false) Only used in mkldnn kernel. Used "
+                  "whenever fc output is as an input to residual connection.")
+        .SetDefault(false)
+        .AsExtra();
```
Comment (tsocha) [inline paddle/fluid/operators/fc_op.cc]: As above.

File: paddle/fluid/framework/ir/graph_traits.cc [hunk 3]
```diff
@@ -23,6 +26,52 @@ namespace ir {
...
+bool IsReachable(ir::Graph *graph, Node *from, Node *to) {
+  auto find_node = [](ir::Graph *graph, const Node *node) -> Node * {
+    for (auto n : graph->Nodes()) {
+      if (n == node) {
+        return n;
+      }
+    }
+
+    return nullptr;
+  };
```
Comment (tsocha) [inline paddle/fluid/framework/ir/graph_traits.cc]: Could you make this lambda a normal function?

File: python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_fc_elementwise_add_fuse_pass.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+import hypothesis
+from hypothesis import given, settings, seed, example, assume
+import hypothesis.strategies as st
+
+
+class TestFCElementwiseAddMkldnnFusePass(PassAutoScanTest):
+    def sample_program_config(self, draw):
+        axis = draw(st.sampled_from([-1, 0, 1]))
+        FCAsX = draw(st.sampled_from([True, False]))
```
Comment (tsocha) [inline python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_fc_elementwise_add_fuse_pass.py:31]: Make it more descriptive, please.
```suggestion
        FCAsInput = draw(st.sampled_from([True, False]))
```

File: python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_fc_elementwise_add_fuse_pass.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            },
+            outputs={"Out": ["fc_output"]},
+            attrs={
+                "use_mkldnn": True,
+                "padding_weights": False,
+                "activation_type": "",
+                "in_num_col_dims": 1,
+            })
+
+        if FCAsX:
```
Comment (tsocha) [inline python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_fc_elementwise_add_fuse_pass.py:65]:
```suggestion
        if FCAsInput:
```

File: paddle/fluid/operators/mkldnn/fc_mkldnn_op.cc [hunk 6]
```diff
@@ -406,36 +406,39 @@ class FCPrimitiveFactory {
...
-        inner_scale = ctx.Attr<float>("Scale_out");
-      }
-    }
+    float activation_scale =
+        force_fp32_output ? 1.0f : has_activation ? ctx.Attr<float>("Scale_out")
+                                                  : 1.0f;
+    float scale_out_data =
+        force_fp32_output ? 1.0f : has_activation
+                                       ? 1.0f
+                                       : ctx.Attr<float>("Scale_out");
```
Comment (tsocha) [inline paddle/fluid/operators/mkldnn/fc_mkldnn_op.cc]: I can see 1.0f value in both places for scale.
Please try to merge both conditions into one.

File: paddle/fluid/operators/mkldnn/fc_mkldnn_op.cc [hunk 7]
```diff
@@ -406,36 +406,39 @@ class FCPrimitiveFactory {
   // scaled with its own scales, this data needs to be divided by
   // those scales to normalise them back to what their floating-point range
   // was. Then we multiply them by desired output scale we want on the output.
-  std::tuple<std::vector<float>, float> ComputeOutputShiftScale(
+  std::tuple<float, std::vector<float>, float> ComputeOutputShiftScale(
```
Comment (tsocha) [inline paddle/fluid/operators/mkldnn/fc_mkldnn_op.cc]: Why did you add a new tuple element at the beginning instead of the end?

File: paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.h [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+#include <boost/optional.hpp>
+#include <memory>
+#include <string>
+#include <tuple>
+#include <utility>
+
+#include "paddle/fluid/framework/ir/fuse_pass_base.h"
+#include "paddle/fluid/framework/ir/graph.h"
+#include "paddle/fluid/framework/ir/graph_pattern_detector.h"
```
Comment (tsocha) [inline paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.h]: Do we need all of these includes in header?
Please verify if its needed or we can move it into cpp file/remove it.

File: paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+      .End()
+      .AddOutput("Out")
+      .IsTensor()
+      .End()
+      .AddAttr("axis")
+      .IsIntIn({-1, 0, 1})
+      .End();
+}
+
+GraphWithStats FCResidualConnectionMKLDNNFusePass::FuseFCAsX(
```
Comment (tsocha) [inline paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc]: FuseFCAsX is too similar to FuseFCAsY
Please keep it DRY.
image

File: paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+      projection_node = fc_x_output;
+      residual_fc_op = fc_y_op;
+      residual_fc_output = fc_y_output;
+    } else if (IsReachable(g, fc_y_input, fc_x_output)) {
+      projection_node = fc_y_output;
+      residual_fc_op = fc_x_op;
+      residual_fc_output = fc_x_output;
+    } else {
+      return;
+    }
```
Comment (tsocha) [inline paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc:273]: I can see that many lines of code is duplicated.
Since `A+B == B+A` find a way to reduce it, please.
Maybe make a pass which will enforce that `FC` is always the `X input` for `Add`?

**Review by tsocha — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_fc_elementwise_add_fuse_pass.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+                        "Y": ["input_data"]},
+                outputs={"Out": ["elementwise_output"]},
+                attrs={'axis': axis})
+        else:
+            elt_add_op = OpConfig(
+                type="elementwise_add",
+                inputs={"X": ["input_data"],
+                        "Y": ["fc_output"]},
+                outputs={"Out": ["elementwise_output"]},
+                attrs={'axis': axis})
```
Comment (tsocha) [inline python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_fc_elementwise_add_fuse_pass.py]:
```suggestion
        if FCAsX:
            inputs={"X": ["fc_output"],
                    "Y": ["input_data"]},
        else:
            inputs={"X": ["input_data"],
                    "Y": ["fc_output"]},

        elt_add_op = OpConfig(
            type="elementwise_add",
            inputs=inputs
            outputs={"Out": ["elementwise_output"]},
            attrs={'axis': axis})
```

**Review by tsocha — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/mkldnn/fc_mkldnn_op.cc (hunk 7 above)
Comment (Silv3S) [reply paddle/fluid/operators/mkldnn/fc_mkldnn_op.cc]: I wanted to follow the existing pattern because similar quantization is done for Conv operator.

Anyway, I delete all int8 changes (because they are not finished) from this PR to make it more concise.

File: paddle/fluid/operators/fc_op.cc (hunk 1 above)
Comment (Silv3S) [reply paddle/fluid/operators/fc_op.cc]: Thank you for pointing this out. I was sure that we need to declare input/attributes in base op before they can be referenced in other ops. Now I removed them from base op and everything works the same

File: python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_fc_elementwise_add_fuse_pass.py [hunk 12]
```diff
@@ -56,26 +56,22 @@ def generate_fc_bias():
             },
             outputs={"Out": ["fc_output"]},
             attrs={
-                "use_mkldnn": True,
+                "use_mkldnn": False,
```
Comment (tsocha) [inline python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_fc_elementwise_add_fuse_pass.py]: Are you sure this attribute should be False?

File: paddle/fluid/inference/api/paddle_pass_builder.cc [hunk 13]
```diff
@@ -294,6 +294,7 @@ void CpuPassStrategy::EnableMKLDNN() {
              // Disabled due to topology-dependent speed-up
              //  "fc_mkldnn_pass",
              //  "fc_act_mkldnn_fuse_pass",
```
Comment (tsocha) [inline paddle/fluid/inference/api/paddle_pass_builder.cc:305]: Please verify if we can turn these passes on.

Comment (lidanqing-vv) [reply paddle/fluid/inference/api/paddle_pass_builder.cc:305]: @Silv3S Hi, could you please add a VIT/OCR test in the API test. And use insert_pass, to enable the fc_mkldnn_pass and fc_act_mkldnn_fuse_pass? I will find you reference

Comment (lidanqing-vv) [reply paddle/fluid/inference/api/paddle_pass_builder.cc:305]: In https://github.com/PaddlePaddle/Paddle-Inference-Demo/blob/master/python/x86_lstm_demo/model_test.py
```
    config.switch_ir_optim(True)
    # This pass must be added before fc_fuse_pass to work properly
    config.pass_builder().insert_pass(5, "fc_lstm_fuse_pass")
    config.pass_builder().append_pass("fc_mkldnn_pass")
```

Comment (lidanqing-vv) [reply paddle/fluid/inference/api/paddle_pass_builder.cc:305]: @Silv3S  You may need to count together with passes inside `CpuPassStrategy()`, because those are pushed firstly into `pass_` and then mkldnn passes followed. 

You can verify if it is correct by running and observing the execution log.

The fc related two passes are currently commented now, because in some models with the fc passes, fp32 oneDNN is slower than fp32 original. We need to figure out a good way later.

File: python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_fc_elementwise_add_fuse_pass.py (hunk 12 above)
Comment (Silv3S) [reply python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_fc_elementwise_add_fuse_pass.py]: I tested both scenarios:
- if `fc_mkldnn_op` with given attributes is fused correctly,
- if base `fc_op` is fused correctly with sequence of `fc_mkldnn_pass` and `fc_elementwise_add_mkldnn_fuse_pass`.

Anyway, now I think that it's unnecessary and I will change it back to `use_mkldnn: True`

**Comment by lidanqing-vv** [comment]
@Silv3S Hi, please write in PR description how much percentage this PR speed up VIT-OCR, this will speedup PR merging by Baidu. Thank you.

File: paddle/fluid/inference/api/paddle_pass_builder.cc (hunk 13 above)
Comment (Silv3S) [reply paddle/fluid/inference/api/paddle_pass_builder.cc:305]: Thank you for providing examples. I will use `GetPassIndex()` method from https://github.com/PaddlePaddle/Paddle/pull/38643 to dynamically insert these two passes before `fc_elementwise_add_mkldnn_fuse_pass`.

File: paddle/fluid/inference/api/paddle_pass_builder.h [hunk 14]
```diff
@@ -71,6 +71,10 @@ class PD_INFER_DECL PaddlePassBuilder {
   /// \param[in] idx the position to delete.
   void DeletePass(size_t idx);
 
+/// \brief Get the certain position of a pass.
+  /// \param[in] pass_type the type of insert pass.
```
Comment (tsocha) [inline paddle/fluid/inference/api/paddle_pass_builder.h]:
```suggestion
  /// \brief Get the certain position of a pass.
  /// \param[in] pass_type the type of insert pass.
```

File: paddle/fluid/inference/tests/api/analyzer_vit_ocr_tester.cc [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace paddle {
+namespace inference {
+namespace analysis {
+
+void SetConfig(AnalysisConfig *cfg, bool use_mkldnn = false) {
+  cfg->SetModel(FLAGS_infer_model + "/inference.pdmodel",
+                FLAGS_infer_model + "/inference.pdiparams");
+
+  cfg->DisableGpu();
```
Comment (tsocha) [inline paddle/fluid/inference/tests/api/analyzer_vit_ocr_tester.cc]: Do we want to disable GPU even when OneDNN is disabled?

File: paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+      .IsTensor()
+      .End()
+      .AddAttr("axis")
+      .IsIntIn({-1, 0, 1})
+      .End();
+}
+
+GraphWithStats FCResidualConnectionMKLDNNFusePass::FuseFC(
+    const std::string& name_scope, const GraphWithStats& graph_with_stats,
+    bool asX) const {
```
Comment (tsocha) [inline paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc]:
```suggestion
GraphWithStats FCResidualConnectionMKLDNNFusePass::FuseFC(
    const std::string& name_scope, const GraphWithStats& graph_with_stats,
    bool FCasX) const {
```
Lets rename it to provide information `What is as X`.

File: paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+  GraphPatternDetector gpd;
+  auto pattern = gpd.mutable_pattern();
+  patterns::FCMKLDNN fc_pattern{pattern, name_scope};
+  auto fc_output = fc_pattern(
+      gpd.mutable_pattern()->NewNode("fc")->AsInput()->assert_is_op_input(
+          "fc", "Input"),
+      true);
```
Comment (tsocha) [inline paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc]:
```suggestion
  auto fc_output = fc_pattern(
      gpd.mutable_pattern()->NewNode("fc")->AsInput()->assert_is_op_input(
          "fc", "Input"),
      FCAsX);
```

File: paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+    GET_IR_NODE_FROM_SUBGRAPH(residual_data, residual_data,
+                              elementwise_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(elementwise_out, elementwise_out,
+                              elementwise_pattern);
+
+    if (FindFuseOption(*fc_op, *elementwise_op) != FUSE_MKLDNN) return;
+
+    if (!IsReachable(g, residual_data, fc_output)) return;
+
+    if (HasFusedActivation(fc_op)) return;
```
Comment (tsocha) [inline paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc]: I don't think we need here so many blank lines.
```suggestion
    if (FindFuseOption(*fc_op, *elementwise_op) != FUSE_MKLDNN) return;
    if (!IsReachable(g, residual_data, fc_output)) return;
    if (HasFusedActivation(fc_op)) return;
```

File: paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+
+    if (!IsCompat(subgraph, g)) {
+      LOG(WARNING)
+          << "fc_elementwise_add_mkldnn_fuse_pass in op compat failed.";
```
Comment (tsocha) [inline paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc]:
```suggestion
          << "op compat for fc_elementwise_add_mkldnn_fuse_pass failed.";
```

File: paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.h [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+namespace paddle {
+namespace framework {
+namespace ir {
+
+using GraphWithStats = std::pair<ir::Graph*, int>;
+
+class FCResidualConnectionMKLDNNFusePass : public FusePassBase {
+ private:
+  GraphWithStats FuseFC(const std::string& name_scope,
+                           const GraphWithStats& graph_with_stats, bool asX) const;
```
Comment (tsocha) [inline paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.h]:
```suggestion
                           const GraphWithStats& graph_with_stats, bool FCasX) const;
```

**Review by tsocha — DISMISSED** [review]

File: paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc (hunk 17 above)
Comment (Silv3S) [reply paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc]: There are two pattern used, one for **FC** and one for **elementwise_add**
- `as_x` flag is used in `ResidualElementwise` pattern,
- `true` here is for specifying if `FCMKLDNN` pattern has bias

**Review by jakpiase — DISMISSED** [review]
I have really tried to find some major issues in this PR, but found none! Really god job, LGTM

File: paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc (hunk 17 above)
Comment (tsocha) [reply paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc]: Please document this flag with a helper variable
```suggestion
  bool fc_has_bias = true;
  auto fc_output = fc_pattern(
      gpd.mutable_pattern()->NewNode("fc")->AsInput()->assert_is_op_input(
          "fc", "Input"),
      fc_has_bias);
```

**Review by tsocha — CHANGES_REQUESTED** [review]

**Review by tsocha — DISMISSED** [review]

**Review by jczaja — DISMISSED** [review]
LGTM

**Comment by jczaja** [comment]
@zhiqiu Could you please review changes at fc_mkldnn_op.cc related to ShareDataWith ?   Changes are modeled after conv_mkldnn_op.cc  where residual connections part is also using ShareDataWith(). We tried to use TensorCopy but there was significant performance degradation, so FC and CONV both are using ShareDataWith. We need your approval to pass CI. Feel free to ask questions

**Comment by jczaja** [comment]
@baoachun Please review

**Comment by baoachun** [comment]
Hi @Silv3S, well done！Could you please attach performance and accuracy comparison data for this pass?

**Comment by baoachun** [comment]
Hi @Silv3S, well done！Could you please attach performance and accuracy comparison data for this pass?

**Comment by lidanqing-vv** [comment]
@baoachun After @Silv3S provide performance, could you please help with PR-CI-APPROVAL? Thanks !
From the PR description, I can see `with this fuse pass VIT-OCR is performing 4% faster`.

**Comment by zhiqiu** [comment]
> @zhiqiu Could you please review changes at fc_mkldnn_op.cc related to ShareDataWith ? Changes are modeled after conv_mkldnn_op.cc where residual connections part is also using ShareDataWith(). We tried to use TensorCopy but there was significant performance degradation, so FC and CONV both are using ShareDataWith. We need your approval to pass CI. Feel free to ask questions

Hi, it is not suggested to `ShareData` of the input and output of an operator, since it makes the op be an implicit `in-place` indeed. So is it possible to make `residual` the same varariable as `Output`, just like `Param` and `ParamOut` in `sgd` operator.

**Comment by Silv3S** [comment]
> Hi @Silv3S, well done！Could you please attach performance and accuracy comparison data for this pass?

Hi @baoachun, thank you! I updated PR description to provide performance comparison for affected models. I also prepared test which compares results between NativeConfig and AnalysisConfig, but I will need your help with uploading needed data.txt file to vit-ocr.tgz repository.

**Comment by Silv3S** [comment]
> Hi, it is not suggested to `ShareData` of the input and output of an operator, since it makes the op be an implicit `in-place` indeed. So is it possible to make `residual` the same varariable as `Output`, just like `Param` and `ParamOut` in `sgd` operator.

Thank you for review @zhiqiu. Since we don't modify `residual_param` inside `FC_op`, I set it explicitly as output. Now `ShareDataWith` is called on two outputs of operator. Please advise if this approach is good.

**Review by zhiqiu — DISMISSED** [review]
LGTM

**Comment by zhiqiu** [comment]
> > Hi, it is not suggested to `ShareData` of the input and output of an operator, since it makes the op be an implicit `in-place` indeed. So is it possible to make `residual` the same varariable as `Output`, just like `Param` and `ParamOut` in `sgd` operator.
> 
> Thank you for review @zhiqiu. Since we don't modify `residual_param` inside `FC_op`, I set it explicitly as output. Now `ShareDataWith` is called on two outputs of operator. Please advise if this approach is good.

Thanks, LGTM

**Comment by Silv3S** [comment]
Hi @XieYunshen, could you please approve setting timeout for new unit test?

**Comment by Silv3S** [comment]
Continued in https://github.com/PaddlePaddle/Paddle/pull/41776 due to CI bug (non refreshing checks). Please approve second PR (code didn't change)
